src/main/scala/workflow/graph/EquivalentNodeMergeRule.scala

```diff
@@ -0,0 +1,47 @@
+package workflow.graph
+
+/**
+ * A rule to merge equivalent nodes in the DAG.
+ * Nodes are considered equivalent if:
+ * - The operators stored within the nodes are equal (.equals() is true)
+ * - They share the same dependencies
+ *
+ * This rule also merges prefixes if any of
+ * the nodes being merged have their prefix attached.
+ */
+object EquivalentNodeMergeRule extends Rule {
+  override def apply(plan: Graph, prefixes: Map[NodeId, Prefix]): (Graph, Map[NodeId, Prefix]) = {
+    val nodeSetsToMerge = plan.nodes.groupBy(id => (plan.getOperator(id), plan.getDependencies(id))).values
+
+    if (nodeSetsToMerge.size == plan.nodes.size) {
+      // no nodes are mergable
+      (plan, prefixes)
+    } else {
+      nodeSetsToMerge.filter(_.size > 1).foldLeft((plan, prefixes)) {
+        case ((curPlan, curPrefixes), setToMerge) => {
+          // Construct a graph that merges all of the nodes
+          val nodeToKeep = setToMerge.minBy(_.id)
+          val nextGraph = (setToMerge - nodeToKeep).foldLeft(curPlan) {
+            case (partialMergedPlan, nodeToMerge) => {
+              partialMergedPlan
+                .replaceDependency(nodeToMerge, nodeToKeep)
+                .removeNode(nodeToMerge)
+            }
+          }
+
+          // If any of the nodes being merged have been executed, update the prefixes
```

Comment:
I am having a bit of a tough time following why this node should need to know about prefixes, or why prefixes should be something considered after the first stage of optimization at all? 

If we combine the first two rules into a single rule, and eliminate "prefixes" from the `Rule` interface, do we get to kill this code?

Are we giving anything up other than chaining other rules that might need prefixes (which don't yet exist)? Or is there something I'm not considering with respect to carrying results of optimization forward or something?